versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreIndexImpl.java

```diff
@@ -0,0 +1,535 @@
+ *   <li>Cross check whether the left-truncation used in the serialized representation of this
+ *       implementation is really legit in real life. <em>It still feels valid and legit and
+ *       efficient.</em>
+ *   <li>Add some checksum (e.g. {@link java.util.zip.CRC32C}, preferred, or {@link
+ *       java.util.zip.CRC32}) to the serialized representation?
+ * </ul>
+ */
+final class StoreIndexImpl<V> implements StoreIndex<V> {
+
+  static final int MAX_KEY_BYTES = 4096;
+
+  /**
+   * Assume 4 additional bytes for each added entry: 2 bytes for the "strip" and 2 bytes for the
+   * "add" var-ints.
+   */
+  private static final int ASSUMED_PER_ENTRY_OVERHEAD = 2 + 2;
+
+  public static final Comparator<StoreIndexElement<?>> KEY_COMPARATOR =
+      Comparator.comparing(StoreIndexElement::key);
+
+  private final int deserializedSize;
+  private int estimatedSerializedSizeDiff;
+  private final List<StoreIndexElement<V>> elements;
+  private final ElementSerializer<V> serializer;
+
+  private boolean modified;
+  private ObjId objId;
+
+  // NOTE: The implementation uses j.u.ArrayList to optimize for reads. Additions to this data
+  // structure are rather inefficient, when elements need to be added "in the middle" of the
+  // 'elements' j.u.ArrayList.
+
+  StoreIndexImpl(ElementSerializer<V> serializer) {
+    this(new ArrayList<>(), 1, serializer, false);
+  }
+
+  StoreIndexImpl(
+      List<StoreIndexElement<V>> elements,
+      int deserializedSize,
+      ElementSerializer<V> serializer,
+      boolean modified) {
+    this.elements = elements;
+    this.deserializedSize = deserializedSize;
+    this.serializer = serializer;
+    this.modified = modified;
+  }
+
+  @Override
+  public boolean isModified() {
+    return modified;
+  }
+
+  @Override
+  public ObjId getObjId() {
+    return objId;
+  }
+
+  @Override
+  public StoreIndex<V> setObjId(ObjId objId) {
+    this.objId = objId;
+    return this;
+  }
+
+  @Override
+  public StoreIndex<V> loadIfNecessary(Set<StoreKey> keys) {
+    return this;
+  }
+
+  @Override
+  public boolean isLoaded() {
+    return true;
+  }
+
+  @Override
+  public StoreIndex<V> asMutableIndex() {
+    return this;
+  }
+
+  @Override
+  public boolean isMutable() {
+    return true;
+  }
+
+  @Override
+  public List<StoreIndex<V>> divide(int parts) {
+    List<StoreIndexElement<V>> elems = elements;
+    int size = elems.size();
+    checkArgument(
+        parts > 0 && parts <= size,
+        "Number of parts %s must be greater than 0 and less or equal to number of elements %s",
+        parts,
+        size);
+    int partSize = size / parts;
+    int serializedMax = deserializedSize + estimatedSerializedSizeDiff;
+
+    List<StoreIndex<V>> result = new ArrayList<>(parts);
+    int index = 0;
+    for (int i = 0; i < parts; i++) {
+      int end = i < parts - 1 ? index + partSize : elems.size();
+      List<StoreIndexElement<V>> partElements = new ArrayList<>(elements.subList(index, end));
```

Comment:
The real serialized size can be completely off (due to the key-prefix-encoding), so it's pretty hard to do that.
But I hope the code / object-format is flexible enough to allow such changes in the future.

Reply:
(maybe for follow-up). As far as I understand index splitting may occur only after the stripes have been serialized. Since Deserializaion (for splitting) follows the natural key order, we know the side "before" each key. Splitting also follows the natural key order, so we could use the size marks from the (de-)serialization step for splitting by size (I hope I did not miss anything :) )